Fix reproducibility issues

mzf-guest opened the pull request:
Hello,

here the patch from the Debian package that fixes reproducibility issues:

 - Replace `date` by [SOURCE_DATE_EPOCH](https://reproducible-builds.org/docs/source-date-epoch/)
 - Fix file locale dependent sort of `ls` for [stable inputs](https://reproducible-builds.org/docs/stable-inputs/)

(This pull request invalidates the [previous one](https://github.com/FreeFem/FreeFem-sources/pull/186), based on master instead of develop. Sorry for the noise.)

Best,

François

prj- commented on examples/eigen/Makefile.am:
```diff
@@ -66,4 +66,4 @@ all.edp:
 	echo includepath = \"../../idp/\" > freefem++.pref
-	echo loadpath = \"../../plugin/seq/\" >> freefem++.pref
\ No newline at end of file
+	echo loadpath = \"../../plugin/seq/\" >> freefem++.pref
```
Please revert this change.

mzf-guest replied:
Done!

prj- commented:
@mzf-guest, should we close this, or could you address my prior point?

mzf-guest commented:
@prj- I've just updated the problematic file to remove the tab to white-space modification. I've also rebased to recent `develop`branch, so it should be ready to merge. If not, just let me know and I'll update accordingly.

Thanks!

prj- commented:
Sorry, I just realised that my comment back from August 2021 was pending and that I did not submit my review...

prj- commented on configure.ac:
```diff
@@ -185,7 +185,9 @@ m4_define([TOOL_DISABLE_NO],
 ])
 
-echo "#  Build with FreeFEM with ./configure " `date`  >$ff_where_lib_conf
+BUILD_DATE="$(date --utc --date="@${SOURCE_DATE_EPOCH:-$(date +%s)}" +%Y-%m-%d)"
```
```suggestion
DATE_FMT="+%Y-%m-%d"
SOURCE_DATE_EPOCH="${SOURCE_DATE_EPOCH:-$(date +%s)}"
BUILD_DATE=$(date -u -d "@$SOURCE_DATE_EPOCH" "$DATE_FMT" 2>/dev/null || date -u -r "$SOURCE_DATE_EPOCH" "$DATE_FMT" 2>/dev/null || date -u "$DATE_FMT")```

prj- commented on configure.ac:
````diff
@@ -185,7 +185,11 @@ m4_define([TOOL_DISABLE_NO],
-echo "#  Build with FreeFEM with ./configure " `date`  >$ff_where_lib_conf
+DATE_FMT="+%Y-%m-%d"
+SOURCE_DATE_EPOCH="${SOURCE_DATE_EPOCH:-$(date +%s)}"
+BUILD_DATE=$(date -u -d "@$SOURCE_DATE_EPOCH" "$DATE_FMT" 2>/dev/null || date -u -r "$SOURCE_DATE_EPOCH" "$DATE_FMT" 2>/dev/null || date -u "$DATE_FMT")```
````
```suggestion
BUILD_DATE=$(date -u -d "@$SOURCE_DATE_EPOCH" "$DATE_FMT" 2>/dev/null || date -u -r "$SOURCE_DATE_EPOCH" "$DATE_FMT" 2>/dev/null || date -u "$DATE_FMT")
```
Sorry about that...

prj- commented on src/fflib/Makefile.am:
```diff
@@ -105,6 +105,7 @@ endif
+BUILD_DATE="$(date --utc --date="@${SOURCE_DATE_EPOCH:-$(date +%s)}" +%Y-%m-%d)"
 strversionnumber.cpp: $(libff_a_SOURCES2) ../../Makefile 
-	m4 -DVersionFreeFemDate="`date`" -DGitVersion="`git describe --tags 2>/dev/null|| echo 'no git'`" strversionnumber.m4 > $@
+	m4 -DVersionFreeFemDate="$(BUILD_DATE)" -DGitVersion="`git describe --tags 2>/dev/null|| echo 'no git'`" strversionnumber.m4 > $@
```
This does not work on even on Debian.
```console
$ make strversionnumber.cpp
m4 -DVersionFreeFemDate="""" -DGitVersion="`git describe --tags 2>/dev/null|| echo 'no git'`" strversionnumber.m4 > strversionnumber.cpp
```

Any suggestion @mzf-guest?

mzf-guest replied:
Indeed! This code does not work when `SOURCE_DATE_EPOCH` is not defined by the build chain. Good catch! and sorry the miss.

The Makefile example from [this page](https://reproducible-builds.org/docs/source-date-epoch/) seems to work better, at least in a minimal test. But I can't make it work from the Makefile.am file. When running `autoreconf -i`, the `ifeq` is recognized as a AM command instead of copying it verbatim in the Makefile.

```
DATE_FMT = +%Y-%m-%d
ifdef SOURCE_DATE_EPOCH
    BUILD_DATE ?= $(shell date -u -d "@$(SOURCE_DATE_EPOCH)" "$(DATE_FMT)" 2>/dev/null || date -u -r "$(SOURCE_DATE_EPOCH)" "$(DATE_FMT)" 2>/dev/null || date -u "$(DATE_FMT)")
else
    BUILD_DATE ?= $(shell date "$(DATE_FMT)")
endif
```

Unfortunately, I'm not very skilled with autotools. Do you have any hints?

prj- replied:
Ahah, no, I faced the exact same issue. Maybe this should be put a static makefile, like `Makefile.date`. And then we'd replace `date` by `make -f Makefile.date date` with
```console
$ cat Makefile.date
DATE_FMT = +%Y-%m-%d
ifdef SOURCE_DATE_EPOCH
    BUILD_DATE ?= $(shell date -u -d "@$(SOURCE_DATE_EPOCH)" "$(DATE_FMT)" 2>/dev/null || date -u -r "$(SOURCE_DATE_EPOCH)" "$(DATE_FMT)" 2>/dev/null || date -u "$(DATE_FMT)")
else
    BUILD_DATE ?= $(shell date "$(DATE_FMT)")
endif

date:
    echo $(BUILD_DATE)
```

prj- replied:
Any update about this @mzf-guest?
